client/hwlib/src/collectors/hardware_info.rs

```diff
@@ -0,0 +1,330 @@
+        _ => None,
+    };
+
+    let bios = devices::Bios {
+        firmware_revision,
+        release_date: Some(release_date),
+        revision: Some(bios_info.version().to_string()),
+        vendor: bios_info.vendor().to_string(),
+        version: bios_info.version().to_string(),
+    };
+    Ok(bios)
+}
+
+/// Retrieve CPU information from SMBIOS
+pub fn collect_processor_info_smbios(
+    processor_info: &SMBiosProcessorInformation,
+    max_cpu_frequency_filepath: &'static str,
+) -> Result<devices::Processor> {
+    let cpu_id = CpuId::new(processor_info)?;
+    let cpu_freq = CpuFrequency::from_file(max_cpu_frequency_filepath)
+        .unwrap()
+        .mHz();
+
+    let processor = devices::Processor {
+        codename: cpu_id.codename().unwrap_or_else(|| "Unknown".to_string()),
+        frequency: cpu_freq,
+        manufacturer: processor_info.processor_manufacturer().to_string(),
+        version: processor_info.processor_version().to_string(),
+    };
+    Ok(processor)
+}
+
+/// Retrieve CPU information from cpuinfo file
+pub fn retrieve_processor_info_cpuinfo(
+    cpuinfo_filepath: &'static str,
+    max_cpu_frequency_filepath: &'static str,
+) -> Result<devices::Processor> {
+    let cpu_info = CpuInfo::from_file(cpuinfo_filepath)?;
+    let cpu_freq = CpuFrequency::from_file(max_cpu_frequency_filepath)
+        .unwrap()
+        .mHz();
+    let processor = devices::Processor {
+        codename: String::new(),
+        frequency: cpu_freq,
+        manufacturer: cpu_info.cpu_type,
+        version: cpu_info.model,
+    };
+    Ok(processor)
+}
+
+pub fn collect_chassis_info(
+    chassis_info: &SMBiosSystemChassisInformation,
+) -> Result<devices::Chassis> {
+    let chassis_type = chassis_info
+        .chassis_type()
+        .ok_or("Failed to get chassis type")
+        .unwrap()
+        .to_string();
+    let manufacturer = chassis_info.manufacturer().to_string();
+    let sku = chassis_info.sku_number().to_string();
+    let version = chassis_info.version().to_string();
+
+    let chassis = devices::Chassis {
+        chassis_type,
+        manufacturer,
+        sku,
+        version,
+    };
+
+    Ok(chassis)
+}
+
+pub fn collect_motherboard_info(board_info: &SMBiosBaseboardInformation) -> Result<devices::Board> {
+    let manufacturer = board_info.manufacturer().to_string();
+    let product_name = board_info.product().to_string();
+    let version = board_info.version().to_string();
+
+    let board = devices::Board {
+        manufacturer,
+        product_name,
+        version,
+    };
+
+    Ok(board)
+}
+
+pub fn collect_motherboard_info_from_device_tree(
+    device_tree_filepath: &'static str,
+) -> Result<devices::Board> {
+    let base_path = std::path::Path::new(device_tree_filepath);
+
+    let try_read = |file| {
+        std::fs::read_to_string(base_path.join(file))
+            .map(|s| s.trim().to_string())
+            .unwrap_or_else(|_| "Unknown".to_string())
```

Comment:
If I remove `to_string` from the `s.trim()` call, I get an error:
> returns a value referencing data owned by the current function. `s` is borrowed here

So, I convert the result to a String within the closure itself. Also, if I replace `.unwrap_or_else(|_| "Unknown".to_string())` with `.unwrap_or("Unknown")` without `to_string()` call, I get an error:
> expected `String`, found `&str``

I've managed to improve it the following way:
```rust
        let try_read = |file| {
            fs::read_to_string(base_path.join(file))
                .map(|s| s.trim().to_string())
                .unwrap_or("Unknown".to_string())
        };
```

Reply:
Just a quick note while we're here, taking the `base_path` as a closure upvalue makes the calls later on feel quite odd. I think it would be cleaner to explicitly call `base_path.join` when making the parameter, so this function would just take a path to read and do nothing more

Reply:
Updated with your suggestion